### PR TITLE
Adding heap access to asm->wasm builder.

### PR DESCRIPTION
**Pull request by flagxor**
Refactor assignment slightly.

### CODE REVIEW

**Comment by flagxor** [comment]
This change needs this in v8 before landing:
https://codereview.chromium.org/1405993009/

**Comment by flagxor** [comment]
The v8 side change has landed.

File: src/wasm/asm-wasm-builder.cc [hunk 1]
```diff
@@ -334,8 +328,68 @@ class AsmWasmBuilderImpl : public AstVisitor {
...
+    } else {
+      UNREACHABLE();
+    }
+    current_function_builder_->AppendCode(
+        WasmOpcodes::LoadStoreOpcodeOf(mtype, is_set_op_), false);
+    is_set_op_ = false;
+    current_function_builder_->AppendCode(
+        WasmOpcodes::LoadStoreAccessOf(mtype), false);
+    Literal* value = expr->key()->AsLiteral();
+    if (value) {
```
Comment (titzer) [inline src/wasm/asm-wasm-builder.cc]: Don't you need to multiply the number by the size of the access (i.e. memtype)?

File: src/wasm/asm-wasm-builder.cc [hunk 2]
```diff
@@ -334,8 +328,68 @@ class AsmWasmBuilderImpl : public AstVisitor {
...
+    }
+    BinaryOperation* binop = expr->key()->AsBinaryOperation();
+    if (binop) {
+      DCHECK(Token::SAR == binop->op());
+      Literal* shift = binop->right()->AsLiteral();
+      DCHECK(shift->raw_value()->IsNumber());
+      DCHECK(kAstI32 == TypeOf(shift));
+      int val = static_cast<int>(shift->raw_value()->AsNumber());
+      DCHECK(size == 1 << val);
+      RECURSE(Visit(binop->left()));
```
Comment (titzer) [inline src/wasm/asm-wasm-builder.cc]: I think we also need to do a mask here?

File: src/wasm/asm-wasm-builder.cc (hunk 1 above)
Comment (flagxor) [inline src/wasm/asm-wasm-builder.cc]: Oops, ah yes for constants. Done

File: src/wasm/asm-wasm-builder.cc (hunk 2 above)
Comment (flagxor) [inline src/wasm/asm-wasm-builder.cc]: Ah, yes, sigh, more perf loss, but correct.
One thing wasm will get us round.

File: src/wasm/asm-wasm-builder.cc [hunk 3]
```diff
@@ -308,22 +313,11 @@ class AsmWasmBuilderImpl : public AstVisitor {
...
-        is_set_op_ = true;
-        RECURSE(Visit(expr->target()));
-        is_set_op_ = false;
-        RECURSE(Visit(expr->value()));
-        break;
-      default:
-        UNREACHABLE();
-    }
+    DCHECK(in_function_);
+    is_set_op_ = true;
```
Comment (titzer) [inline src/wasm/asm-wasm-builder.cc]: Do you need to unset is_set_op in between or before returning?

Comment (flagxor) [inline src/wasm/asm-wasm-builder.cc]: Done.

**Comment by flagxor** [comment]
Pushed revised version.
Needs:
https://codereview.chromium.org/1471073003/

**Comment by flagxor** [comment]
Merging this to match the other side.

**Pull request merged**